File: problems/1195-fizz-buzz-multithreaded.java
// Solved 2022-04-14
// Runtime: 6 ms (84.47%)
// Memory Usage: 42.1 MB (83.55%)

import java.util.function.IntConsumer;
import java.util.function.IntPredicate;

class FizzBuzz {
    private int n;
    private int i;

    public FizzBuzz(int n) {
        this.n = n;
        this.i = 1;
    }

    private void run(IntPredicate test, IntConsumer print) throws InterruptedException {
        while (true) {
            while (i <= n && !test.test(i)) {
                wait();
            }
            if (i > n) {
                break;
            }
            print.accept(i);
            ++i;
            notifyAll();
        }
    }

    // printFizz.run() outputs "fizz".
    public synchronized void fizz(Runnable printFizz) throws InterruptedException {
        run(i -> i % 3 == 0 && i % 5 != 0, i -> printFizz.run());
    }

    // printBuzz.run() outputs "buzz".
    public synchronized void buzz(Runnable printBuzz) throws InterruptedException {
        run(i -> i % 3 != 0 && i % 5 == 0, i -> printBuzz.run());
    }

    // printFizzBuzz.run() outputs "fizzbuzz".
    public synchronized void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
        run(i -> i % 15 == 0, i -> printFizzBuzz.run());
    }

    // printNumber.accept(x) outputs "x", where x is an integer.
    public synchronized void number(IntConsumer printNumber) throws InterruptedException {
        run(i -> i % 3 != 0 && i % 5 != 0, printNumber);
    }
}

File: problems/1195-fizz-buzz-multithreaded.cpp
// Solved 2022-04-15
// Runtime: 20 ms (48.87%)
// Memory Usage: 6.6 MB (100.00%)

class FizzBuzz {
   private:
    int n;
    int i;
    mutex m;
    condition_variable cv;

   public:
    FizzBuzz(int n) {
        this->n = n;
        this->i = 1;
    }

    void run(function<bool()> test, function<void()> print) {
        while (true) {
            unique_lock<mutex> lock(m);
            cv.wait(lock, [&] { return i > n || test(); });
            if (i > n) {
                break;
            }
            print();
            ++i;
            lock.unlock();
            cv.notify_all();
        }
    }

    // printFizz() outputs "fizz".
    void fizz(function<void()> printFizz) {
        run([&] { return i % 3 == 0 && i % 5 != 0; }, [&] { printFizz(); });
    }

    // printBuzz() outputs "buzz".
    void buzz(function<void()> printBuzz) {
        run([&] { return i % 3 != 0 && i % 5 == 0; }, [&] { printBuzz(); });
    }

    // printFizzBuzz() outputs "fizzbuzz".
    void fizzbuzz(function<void()> printFizzBuzz) {
        run([&] { return i % 15 == 0; }, [&] { printFizzBuzz(); });
    }

    // printNumber(x) outputs "x", where x is an integer.
    void number(function<void(int)> printNumber) {
        run([&] { return i % 3 != 0 && i % 5 != 0; }, [&] { printNumber(i); });
    }
};
